include/pybind11/pybind11.h

```diff
@@ -1292,6 +1299,15 @@ class module_ : public object {
         PyModule_AddObject(ptr(), name, obj.inc_ref().ptr() /* steals a reference */);
     }
 
+    /** \rst
+        Mark the module as not requiring the GIL in free-threaded Python builds.
+    \endrst */
+    void set_gil_not_used() {
+#ifdef Py_GIL_DISABLED
+        PyUnstable_Module_SetGIL(m_ptr, Py_MOD_GIL_NOT_USED);
```

Comment:
Should we look at switching to multiphase initialization? (if so, not sure how this would correspond)

Reply:
It's not really relevant for free-threading, but it would allow you to better support multiple interpreters.

Reply:
The multi-phase init equivalent is a [`PyModuleDef_Slot`](https://docs.python.org/3/c-api/module.html#c.PyModuleDef.m_slots) with `{Py_mod_gil, Py_MOD_GIL_NOT_USED}`. For example:

https://github.com/python/cpython/blob/df93f5d4bf9d70036d485666d4dd4f009d37f8b9/Modules/_bisectmodule.c#L465

Reply:
My thought is the API, I'm not sure if `m.set_gil_not_used()` would translate to multiphase initialization?

Reply:
It would not directly translate because you would want to set the flag on the `PyModuleDef` instead of the `PyModuleObject`, so you'd probably want to support it in `create_extension_module()`.

Reply:
I'm thinking we could add options to `create_extension_module`, which then could be passed through `PYBIND11_MODULE`, so this could become `PYBIND11_MODULE(name, m, py::gil_not_used())` for example. This would be compatible with multiphase init.

Reply:
Yeah, I think that would be nicer. I'd imagine you'd want the new argument to be optional -- I'm not exactly sure how to do that with macros. Maybe something like:

```
#define PYBIND11_MODULE(name, variable, ...)
```

and then do something with `__VA_ARGS__`.

Reply:
Yes, I do that in CLI11: https://github.com/CLIUtils/CLI11/blob/08d840bfbe92c517c279c7527323d0d1d4470e2c/include/CLI/App.hpp#L39-L46.

Hardest part is setting up the `py::gil_not_used()` handling, we do it all over the place though, so should be easy. Since there's only one option currently it is also easy to fake it with two different signatures. I can take a stab at it soon. Is it find to add the slot without adding the init/exec slots? I don't mind using the single phase method internally for now too, I just want the API to support multiphase if we add it eventually.

Reply:
I added a commit that does this, using the simple path for now, template meta programming can wait until we have more than one tag (unless someone wants to volunteer to do it!).

Reply:
(I’ll pull this commit off tomorrow and make a PR to this PR instead)

Reply:
PR at https://github.com/colesbury/pybind11/pull/11. CI needs approval.